compiler/rustc_const_eval/src/interpret/memory.rs

```diff
@@ -1209,21 +1209,28 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
                         throw_ub_custom!(fluent::const_eval_copy_nonoverlapping_overlapping);
                     }
                 }
+            }
 
-                for i in 0..num_copies {
-                    ptr::copy(
-                        src_bytes,
-                        dest_bytes.add((size * i).bytes_usize()), // `Size` multiplication
-                        size.bytes_usize(),
-                    );
+            let size_in_bytes = size.bytes_usize();
+            // For particularly large arrays (where this is perf-sensitive) it's common that
+            // we're writing a single byte repeatedly. So, optimize that case to a memset.
+            if size_in_bytes == 1 && num_copies >= 1 {
+                // SAFETY: `src_bytes` would be read from anyway by copies below (num_copies >= 1).
+                // Since size_in_bytes = 1, then the `init.no_bytes_init()` check above guarantees
+                // that this read at type `u8` is OK -- it must be an initialized byte.
```

Comment:
This comment seems confused. SAFETY comments are about the safety of the interpreter itself. The bytes in question are *always* initialized in that sense. The check above is about whether they are initialized *inside the interpreter state*.

Reply:
Happy to edit it! I didn't obviously find any guarantees that the interpreter always initializes backing allocations before passing them around (seems like at minimum based on profiling that's not happening with the destination buffer, so it would be bad to read that at type u8 here?).

I would have expected that the emulated interpreter state having the bytes initialized implies the interpreter's state also has them initialized.

Reply:
The `u8` buffer backing interpreter allocations is always initialized. It has type `u8` rather than `MaybeUninit<u8>`, so documenting that seems redundant.

Reply:
Profiling seems to be misleading then. I think we are calling `calloc` (via `try_new_zeroed_slice`) which might do the zero-init via efficient copy-on-write tricks or so.

Reply:
Perhaps! I'm 99% sure we're not touching every byte until we hit this code. I'll look into changing some things here then.

Reply:
New allocations get created here:

https://github.com/rust-lang/rust/blob/2ea7a37e1113febac8603729e33fdd94f2738807/compiler/rustc_middle/src/mir/interpret/allocation.rs#L293-L333

That ends up calling this:

https://github.com/rust-lang/rust/blob/2ea7a37e1113febac8603729e33fdd94f2738807/compiler/rustc_middle/src/mir/interpret/allocation.rs#L48-L53

I am 100% sure that all these `u8` are initialized.